### PR TITLE
Guard Recording::writeElement against SIGSEGV on a corrupted metadata tree

### PR DESCRIPTION
**Pull request by jbachorik**
**What does this PR do?**:
Adds a null check and a recursion-depth guard (10) to `Recording::writeElement()`, so a corrupted or dangling `Element*` in the JFR metadata tree produces a logged, truncated dump instead of a SIGSEGV. Each guard trip also increments a dedicated counter (`metadata_tree_null_child` / `metadata_tree_depth_exceeded`), which flows out as a `datadog.ProfilerCounter` JFR event through the existing counter-reporting path — a durable, queryable signal in case this recurs, rather than relying on `stderr` from an embedded native library (rarely captured/monitored). Also fixes a macOS x86_64 build issue in `safeAccess.cpp` (`current_pc` referenced an undefined `context_rip` macro instead of the `DU3_PREFIX`-based `uc_mcontext` accessor already used on aarch64), found while testing this change locally.

**Motivation**:
PROF-15075: a SIGSEGV crash cluster in `Recording::writeElement`, seen on JDK 17/21/25 with dd-trace-java 1.63.0. The production trace (`writeElement` x5 self-recursive <- `writeMetadata` <- `switchChunk` <- `Profiler::dump`) is a plain periodic chunk dump with no profiler restart involved, so what actually corrupts/nulls the `Element*` is not yet confirmed. This guard turns the crash into a safe, logged, telemetry-visible failure in the meantime.

**Additional Notes**:
The true trigger remains open. A core dump or heap-sanitizer run is what's needed to confirm it; the new counters and log lines are meant to catch and localize a recurrence in the meantime, not to explain away the underlying issue.

**How to test the change?**:
- `./gradlew :ddprof-lib:buildDebug -Pskip-gtest` and the `flightRecorder_result_ut` gtest suite pass with this change.
- No automated test reproduces the crash itself, since the trigger isn't known yet; the guard is defense-in-depth.

**For Datadog employees**:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
  credentials of any kind, I've requested a security review (run the `dd:platform-security-review`
  skill, or file a request via the [PSEC review form](https://datadoghq.atlassian.net/jira/software/c/projects/PSEC/forms/form/direct/7861446195161534/37715)).
  `bewaire` also runs automatically on every PR.
- [x] This PR doesn't touch any of that.
- [x] JIRA: [PROF-15075](https://datadoghq.atlassian.net/browse/PROF-15075)

Unsure? Have a question? Request a review!

[PROF-15075]: https://datadoghq.atlassian.net/browse/PROF-15075?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jbachorik** [comment]
@codex review

File: ddprof-lib/src/main/cpp/flightRecorder.cpp [hunk 1]
```diff
@@ -1147,10 +1161,23 @@ void Recording::writeElement(Buffer *buf, const Element *e) {
     buf->putVar64(e->_attributes[i]._value);
   }
 
-  buf->putVar64(e->_children.size());
+  size_t child_count = 0;
   for (size_t i = 0; i < e->_children.size(); i++) {
+    if (e->_children[i] != nullptr) {
+      child_count++;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply ddprof-lib/src/main/cpp/flightRecorder.cpp]: Fixed in 1d5e7394c — `countSerializableChildren()` now applies the same depth-truncation rule the recursive writer uses, so the encoded count can never exceed what's actually serialized.

File: ddprof-lib/src/main/cpp/flightRecorder.cpp [hunk 2]
```diff
@@ -1137,7 +1137,21 @@ void Recording::writeHeader(Buffer *buf) {
...
+    // is rarely captured/monitored, and this guard exists precisely because
+    // we don't yet know what corrupts the tree (PROF-15075) — an unmonitored
+    // log line would let that recur invisibly forever.
+    Counters::increment(METADATA_TREE_DEPTH_EXCEEDED);
+    fprintf(stderr, "[ddprof] [ERROR] writeElement depth limit exceeded, truncating output\n");
+    return;
+  }
+
+  if (e == nullptr) {
+    return;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply ddprof-lib/src/main/cpp/flightRecorder.cpp:1174]: Left as documented scope for now (see the spec's "Open follow-up" section, commit a79acf5e4): the depth guard addresses cycles/excessive recursion, not dangling-pointer corruption, which needs a separate fault-protected-read or ownership fix. Tracking it as a follow-up rather than folding it into this fix.

File: ddprof-lib/src/main/cpp/flightRecorder.cpp [hunk 3]
```diff
@@ -1137,7 +1137,21 @@ void Recording::writeHeader(Buffer *buf) {
   flushIfNeeded(buf);
 }
 
-void Recording::writeElement(Buffer *buf, const Element *e) {
+void Recording::writeElement(Buffer *buf, const Element *e, int depth) {
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply ddprof-lib/src/main/cpp/flightRecorder.cpp:1172]: Added in 1d5e7394c — `ddprof-lib/src/test/cpp/flightRecorder_metadata_ut.cpp` covers the null-child, depth-boundary, and pre-boundary cases directly against `countSerializableChildren()`.

**Comment by jbachorik** [comment]
@codex review

File: ddprof-lib/src/main/cpp/flightRecorder.cpp [hunk 4]
```diff
@@ -1145,7 +1145,43 @@ void Recording::writeHeader(Buffer *buf) {
...
+  // itself malformed).
+  bool truncate_children = depth + 1 > 10;
+
+  size_t child_count = 0;
+  for (size_t i = 0; i < children.size(); i++) {
+    if (children[i] == nullptr) {
+      Counters::increment(METADATA_TREE_NULL_CHILD);
+      fprintf(stderr, "[ddprof] [WARN] writeElement skipping null child at index %zu\n", i);
+    } else if (truncate_children) {
+      Counters::increment(METADATA_TREE_DEPTH_EXCEEDED);
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply ddprof-lib/src/main/cpp/flightRecorder.cpp:1163]: warning will not help. it will either be ignored by the user or will annoy them. the counter is for the telemetry

**Review by rkennke — COMMENTED** [review]
**Depth-exceeded truncation is telemetrized but never logged to stderr — unlike the null-child guard.**

In `Recording::writeElement()` (flightRecorder.cpp), the `if (depth > 10) { ...; fprintf(stderr, "...depth limit exceeded...") }` block at the top of the function is dead code. `writeElement()` is only reachable at depth 0..10: the sole recursive call site only fires when `!truncate_children` (i.e. `depth + 1 <= 10`), so `depth` can never exceed 10 when `writeElement()` is entered. That means this `fprintf` — and the `Counters::increment(METADATA_TREE_DEPTH_EXCEEDED)` alongside it — never execute.

The actual telemetry for depth-truncated children does fire, but from a different place: `countSerializableChildren()`'s `else if (truncate_children)` branch, which runs on every `writeElement()` call and correctly increments `METADATA_TREE_DEPTH_EXCEEDED` once per excluded child. It just has no accompanying `fprintf`, unlike the null-child branch right above it, which does both `Counters::increment(METADATA_TREE_NULL_CHILD)` and a `fprintf(stderr, ...)`.

Net effect: a null-child truncation is both counted and logged; a depth-exceeded truncation is only counted, silently. That's inconsistent with the PR's own stated rationale — that stderr logging is needed as a human-visible backstop alongside the counters, precisely because the tree-corruption trigger is unconfirmed.

Suggested fix: move the `fprintf` into `countSerializableChildren()`'s `truncate_children` branch (next to its existing counter increment), and delete the now-unreachable `if (depth > 10)` block in `writeElement()` entirely, rather than trying to make that block reachable. Making it reachable instead (e.g. by loosening the caller's `truncate_children` guard so recursion continues to depth 11) would double-count the same truncation event (once in `countSerializableChildren()`, again in `writeElement()`) and would reintroduce the exact `child_count`/actual-serialized-children mismatch this PR's own regression tests guard against, since `countSerializableChildren()` would have already excluded that child from the encoded count before `writeElement()` serialized it anyway.

File: ddprof-lib/src/test/cpp/flightRecorder_metadata_ut.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2026, Datadog, Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Regression tests for Recording::countSerializableChildren(), the helper
+// writeElement() uses to decide which children of a JfrMetadata::Element
+// tree get serialized when the tree is corrupted (null children) or
+// unexpectedly deep (cycles / excessive recursion).
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply ddprof-lib/src/test/cpp/flightRecorder_metadata_ut.cpp:9]: Good catch — fixed in 63dfc0115. This file was dropped as an unintended side effect of rebasing this branch onto main; it's unrelated to this PR's scope (the reset() regression suite lives on main via #694) and has been restored verbatim.

**Comment by jbachorik** [comment]
Good catch on the dead code — fixed in 8a8d4ce0d. Moved the `fprintf` into `countSerializableChildren()`'s `truncate_children` branch (next to its existing counter increment) and dropped the unreachable `depth > 10` block in `writeElement()`. Depth-exceeded truncation is now logged the same way the null-child case already was. Both regression suites still pass (`gtestDebug_flightRecorder_metadata_ut`, `gtestDebug_jfrMetadata_ut`).
